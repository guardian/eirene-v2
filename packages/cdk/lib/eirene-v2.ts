import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Bucket } from 'aws-cdk-lib/aws-s3';

export class EireneV2 extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'eirene-v2';

		const bucket = new Bucket(this, 'EireneV2Bucket', {
			bucketName: `gu-eirene-v2-${this.stage.toLowerCase()}`,
		});

		// Lambda function not needed in DEV as it is run locally
		if (this.stage !== 'DEV') {
			const lambda = new GuLambdaFunction(this, 'EireneV2Lambda', {
				app,
				functionName: `update-training-input-${this.stage}`,
				fileName: `${app}.zip`,
				runtime: Runtime.NODEJS_22_X,
				handler: 'index.handler',
				environment: {
					STAGE: this.stage,
					S3_BUCKET: bucket.bucketName,
				},
			});

			bucket.grantReadWrite(lambda);
		}

		const dataAccessRole = new Role(this, 'EireneV2DataAccessRole', {
			assumedBy: new ServicePrincipal('comprehend.amazonaws.com'),
			roleName: `EireneV2DataAccessRole-${this.stage}`,
		});
		bucket.grantRead(dataAccessRole);

		// new CfnFlywheel(this, 'EireneV2Flywheel', {
		// 	flywheelName: `${app}-${this.stage}`,
		// 	dataAccessRoleArn: dataAccessRole.roleArn,
		// 	dataLakeS3Uri: bucket.bucketArn,
		// 	modelType: 'DOCUMENT_CLASSIFIER',
		// 	taskConfig: {
		// 		documentClassificationConfig: {
		// 			mode: 'MULTI_CLASS',
		// 		},
		// 		languageCode: 'en',
		// 	}
		// });
	}
}
