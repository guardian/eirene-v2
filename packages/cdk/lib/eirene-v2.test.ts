import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { EireneV2 } from './eirene-v2';

describe('The EireneV2 stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new EireneV2(app, 'EireneV2', {
			stack: 'discussion',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
