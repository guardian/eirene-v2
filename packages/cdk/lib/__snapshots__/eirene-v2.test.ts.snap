// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EireneV2 stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EireneV2Bucket9AE58493": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "gu-eirene-v2-test",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/eirene-v2",
          },
          {
            "Key": "Stack",
            "Value": "discussion",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "EireneV2DataAccessRoleDE2494B8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "comprehend.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "EireneV2DataAccessRole-TEST",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/eirene-v2",
          },
          {
            "Key": "Stack",
            "Value": "discussion",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EireneV2DataAccessRoleDefaultPolicy82C303D3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "EireneV2Bucket9AE58493",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "EireneV2Bucket9AE58493",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EireneV2DataAccessRoleDefaultPolicy82C303D3",
        "Roles": [
          {
            "Ref": "EireneV2DataAccessRoleDE2494B8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EireneV2LambdaD1C37DBB": {
      "DependsOn": [
        "EireneV2LambdaServiceRoleDefaultPolicy4DE56E1D",
        "EireneV2LambdaServiceRole3072E51E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "discussion/TEST/eirene-v2/eirene-v2.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "eirene-v2",
            "S3_BUCKET": {
              "Ref": "EireneV2Bucket9AE58493",
            },
            "STACK": "discussion",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "update-training-input-TEST",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "EireneV2LambdaServiceRole3072E51E",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "eirene-v2",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/eirene-v2",
          },
          {
            "Key": "Stack",
            "Value": "discussion",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "EireneV2LambdaServiceRole3072E51E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "eirene-v2",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/eirene-v2",
          },
          {
            "Key": "Stack",
            "Value": "discussion",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EireneV2LambdaServiceRoleDefaultPolicy4DE56E1D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/discussion/TEST/eirene-v2/eirene-v2.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/discussion/eirene-v2",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/discussion/eirene-v2/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "EireneV2Bucket9AE58493",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "EireneV2Bucket9AE58493",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EireneV2LambdaServiceRoleDefaultPolicy4DE56E1D",
        "Roles": [
          {
            "Ref": "EireneV2LambdaServiceRole3072E51E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
